Source/CarthageKit/Project.swift

```diff
@@ -1165,6 +1183,173 @@ public final class Project { // swiftlint:disable:this type_body_length
 				return incompatibilities.isEmpty ? .init(value: ()) : .init(error: .invalidResolvedCartfile(incompatibilities))
 			}
 	}
+
+	/// Clones the given submodule into the working directory of its parent
+	/// repository, but without any Git metadata.
+	public func cloneSubmoduleInWorkingDirectory(_ submodule: Submodule, _ workingDirectoryURL: URL) -> SignalProducer<(), CarthageError> {
```

Comment:
I don't think these functions should move here:

1. These are generic functions that shouldn't really be tied to the `Project`
2. Because these were moved, the diff for this PR is really large and almost impossible to review

Can you move them back to the Git file?

Reply:
These functions were moved because the new implementation depends on `repositoryFileURL` which is a private function in Project.swift.

Reply:
Then that should be passed in as a parameter

Reply:
What about `cloneOrFetchDependency`? These functions actually rely on `cloneOrFetchDependency` which uses `cloneOrFetchQueue` and `_projectEventsObserver` which are `Project` properties. I think some kind of queue is definitely needed. So also pass the queue in as a parameter?